tests/common/src/lib.rs

```diff
@@ -16,169 +16,395 @@
+
+        message_ids
+            .into_iter()
+            .map(|id| self.get_response_to(id))
+            .collect()
+    }
+
+    pub fn request_batch<E, M, I, D>(&mut self, requests: I) -> Vec<D>
+    where
+        E: Encode,
+        M: Into<MessageDispatchInfo<E>>,
+        I: IntoIterator<Item = M>,
+        D: Decode,
+    {
+        self.try_request_batch(requests)
+            .into_iter()
+            .map(reply_or_panic)
+            .collect()
+    }
+
+    pub fn get_response_to<M, D>(&mut self, id: M) -> Option<Result<D, Error>>
+    where
+        M: Into<MessageId>,
+        D: Decode,
+    {
+        let id = id.into();
+        self.storage()
+            .log
+            .get()
+            .iter()
+            .find(|message| message.reply.map(|(to, _)| to == id).unwrap_or(false))
+            .map(|message| {
+                let (_, exit_code) = message
+                    .reply
+                    .expect("messages that are not replies get filtered above");
+
+                if exit_code != 0 {
+                    Err(Error::Panic)
+                } else {
+                    D::decode(&mut message.payload.as_ref()).map_err(Error::Decode)
+                }
+            })
+    }
+
+    pub fn storage(&mut self) -> &InMemoryStorage {
+        self.runner_state.get_storage()
+    }
+
+    fn runner(&mut self) -> &mut MemoryRunner {
+        self.runner_state.get_runner()
+    }
+
+    fn next_message_id(&mut self) -> MessageId {
+        let message_id = self.message_id.into();
+        self.message_id += 1;
+        message_id
+    }
+
+    fn next_program_id(&mut self) -> ProgramId {
+        let program_id = self.program_id.into();
+        self.program_id += 1;
+        program_id
+    }
+}
+
+fn reply_or_panic<D: Decode>(response: Option<Result<D, Error>>) -> D {
+    match response.expect("No reply for message") {
+        Ok(reply) => reply,
+        Err(Error::Decode(e)) => panic!("Failed to decode reply: {}", e),
+        Err(Error::Panic) => panic!("Request processing error"),
+    }
+}
+
+impl Default for RunnerContext {
+    fn default() -> Self {
+        Self {
+            runner_state: RunnerState::Uninitialzied,
+            program_id: 1,
+            message_id: 1,
+        }
+    }
+}
+
+enum RunnerState {
+    Runner(MemoryRunner),
+    Storage(InMemoryStorage, Config),
+    Uninitialzied,
+}
+
+impl RunnerState {
+    fn get_runner(&mut self) -> &mut MemoryRunner {
+        if let Self::Runner(runner) = self {
+            runner
+        } else {
+            *self = if let Self::Storage(storage, config) = std::mem::take(self) {
+                Self::Runner(Runner::new(&config, storage))
+            } else {
+                Self::Runner(Runner::new(&Config::default(), InMemoryStorage::default()))
+            };
+
+            match self {
+                RunnerState::Runner(runner) => runner,
+                _ => unreachable!(),
+            }
+        }
```

Comment:
A little bit shorter:

```suggestion
        if let Self::Runner(runner) = self {
            runner
        } else {
            let runner = match std::mem::take(self) {
                Self::Storage(storage, config) => Runner::new(&config, storage),
                _ => Runner::new(&Config::default(), InMemoryStorage::default()),
                // TODO: Consider to derive `Default` for `Runner`
            };
            *self = Self::Runner(runner);
            self.get_runner()
        }
```

Reply:
Good idea, implemented recursive tail call, but pattern of keeping enum variant value in separate variable is awkward in `to_storage` function because tuple has to be used, so I decided to keep the original implementation. match in `to_runner` function is more readable, updated. But subjectively it feels less readable in `to_storage` function where if let was preserved.